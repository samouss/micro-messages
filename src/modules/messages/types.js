// @flow

export type MessageId = string;
export type MessageBody = string;

export type Visibility = 'public' | 'private';

export type VisibilityFilter = Visibility | 'all';

export type Message = {
  id: MessageId,
  date: string,
  body: MessageBody,
  visibility: Visibility,
};

export type MessageInput = {
  body: MessageBody,
  visibility: Visibility,
};

export type ByIdState = { [id: MessageId]: Message };
export type IdsState = Array<MessageId>;

export type MessageState = {
  byId: ByIdState,
  ids: IdsState,
};

export type FetchMessagesRequestAction = { type: 'FETCH_MESSAGES_REQUEST' };
export type FetchMessagesSuccessAction = { type: 'FETCH_MESSAGES_SUCCESS', messages: Array<Message> };
export type FetchMessagesFailureAction = { type: 'FETCH_MESSAGES_FAILURE' };

export type PostMessageRequestAction = { type: 'POST_MESSAGE_REQUEST', message: Message };
export type PostMessageSuccessAction = { type: 'POST_MESSAGE_SUCCESS', message: Message };
export type PostMessageFailureAction = { type: 'POST_MESSAGE_FAILURE' };

export type MessageAction =
  | FetchMessagesRequestAction
  | FetchMessagesSuccessAction
  | FetchMessagesFailureAction

  | PostMessageRequestAction
  | PostMessageSuccessAction
  | PostMessageFailureAction;
