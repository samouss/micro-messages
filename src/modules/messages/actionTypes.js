// @flow

export const FETCH_MESSAGES_REQUEST = 'FETCH_MESSAGES_REQUEST';
export const FETCH_MESSAGES_SUCCESS = 'FETCH_MESSAGES_SUCCESS';
export const FETCH_MESSAGES_FAILURE = 'FETCH_MESSAGES_FAILURE';

export const POST_MESSAGE_REQUEST = 'POST_MESSAGE_REQUEST';
export const POST_MESSAGE_SUCCESS = 'POST_MESSAGE_SUCCESS';
export const POST_MESSAGE_FAILURE = 'POST_MESSAGE_FAILURE';
