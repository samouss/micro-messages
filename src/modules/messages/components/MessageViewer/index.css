.MessageViewer {
  flex: 1;
  padding: 25px;
}

.MessageViewer__Header {
  padding-bottom: 25px;
  margin-bottom: 25px;
  border-bottom: 1px solid #E5E5E5;
}

.MessageViewer__Row {
  display: flex;
  margin-bottom: 10px;
}

.MessageViewer__Row:last-child {
  margin-bottom: 0;
}

.MessageViewer__Row__Label {
  margin-right: 10px;
}

.MessageViewer__Content {

}
