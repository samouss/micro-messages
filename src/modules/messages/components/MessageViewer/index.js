// @flow

import React from 'react';
import TimeFromNow from 'components/TimeFromNow';
import Label from 'components/Label';
import Paragraphs from 'components/Paragraphs';
import './index.css';

import type { Message } from '../../types';

export type Props = {
  message: Message,
};

const MessageViewer = ({ message }: Props) => (
  <div styleName="MessageViewer">
    <div styleName="MessageViewer__Header">
      <p styleName="MessageViewer__Row">
        <span styleName="MessageViewer__Row__Label">Reçu le:</span>
        <TimeFromNow
          date={message.date}
        />
      </p>

      <p styleName="MessageViewer__Row">
        <span styleName="MessageViewer__Row__Label">Visibilité:</span>
        <Label
          type={message.visibility}
        />
      </p>
    </div>

    <div styleName="MessageViewer__Content">
      <Paragraphs
        text={message.body}
      />
    </div>
  </div>
);

export default MessageViewer;
