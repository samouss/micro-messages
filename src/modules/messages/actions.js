// @flow

import { fetchMessages as fetchMessagesFromApi } from 'core/api';
import type { Dispatch } from 'store/types';
import * as ACTION_TYPES from './actionTypes';
import type { Message, MessageAction } from './types';

const fetchMessagesRequest = (): MessageAction => ({
  type: ACTION_TYPES.FETCH_MESSAGES_REQUEST,
});

const fetchMessagesSuccess = (messages: Array<Message>): MessageAction => ({
  type: ACTION_TYPES.FETCH_MESSAGES_SUCCESS,
  messages,
});

const fetchMessagesFailure = (): MessageAction => ({
  type: ACTION_TYPES.FETCH_MESSAGES_FAILURE,
});

export const fetchMessages = () => (dispatch: Dispatch) => {
  dispatch(fetchMessagesRequest());

  return fetchMessagesFromApi().then(res => {
    dispatch(fetchMessagesSuccess(res));
  }).catch(() => {
    dispatch(fetchMessagesFailure());
  });
};
