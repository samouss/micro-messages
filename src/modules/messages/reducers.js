// @flow

import { combineReducers } from 'redux';
import * as ACTION_TYPES from './actionTypes';
import type { State, Action } from 'store/types';
import type { MessageId, ByIdState, IdsState } from './types';

const byIdInitialState: ByIdState = {};
const byId = (state = byIdInitialState, action: Action) => {
  switch (action.type) {
    case ACTION_TYPES.FETCH_MESSAGES_SUCCESS: {
      return {
        ...state,
        ...action.messages.reduce((acc, message) => ({
          ...acc,
          [message.id]: message,
        }), {}),
      };
    }

    default: {
      return state;
    }
  }
};

const idsInitialState: IdsState = [];
const ids = (state = idsInitialState, action: Action) => {
  switch (action.type) {
    case ACTION_TYPES.FETCH_MESSAGES_SUCCESS: {
      return state.concat(action.messages.map(message => message.id));
    }

    default: {
      return state;
    }
  }
};

export default combineReducers({
  byId,
  ids,
});

const getState = (state: State) => state.messages;
const getMessageIds = (state: State) => getState(state).ids;

export const getMessage = (state: State, id: MessageId) => {
  return getState(state).byId[id];
};

export const getMessages = (state: State) => {
  return getMessageIds(state).map(id => getMessage(state, id));
};
