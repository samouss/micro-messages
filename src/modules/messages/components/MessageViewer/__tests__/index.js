// @flow

import React from 'react';
import { shallow } from 'enzyme';
import { createMockMessage } from 'test/messages';
import MessageViewer from '../index';

import type { Props } from '../index';

describe('<MessageViewer />', () => {
  it('expect to render the component correctly', () => {
    const props: Props = {
      message: createMockMessage('ID1234'),
    };

    const component = shallow(
      <MessageViewer
        {...props}
      />,
    );

    expect(component).toMatchSnapshot();
  });
});
