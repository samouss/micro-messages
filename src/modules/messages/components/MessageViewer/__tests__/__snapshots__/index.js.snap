// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<MessageViewer /> expect to render the component correctly 1`] = `
<div
  styleName="MessageViewer"
>
  <div
    styleName="MessageViewer__Header"
  >
    <p
      styleName="MessageViewer__Row"
    >
      <span
        styleName="MessageViewer__Row__Label"
      >
        Reçu le:
      </span>
      <TimeFromDate
        date="SOME_DATE"
      />
    </p>
    <p
      styleName="MessageViewer__Row"
    >
      <span
        styleName="MessageViewer__Row__Label"
      >
        Visibilité:
      </span>
      <Label
        type="private"
      />
    </p>
  </div>
  <div
    styleName="MessageViewer__Content"
  >
    <Paragraphs
      text="BODY"
    />
  </div>
</div>
`;
